### PR TITLE
Refactored encoder

### PR DESCRIPTION
**Pull request by anjaligopi**
Encoder v2 is the re-written encoder (specaugment on non-input stacking).
Added support for WPM, specaugment, CNN, bidi LSTMs - refactored librispeech_ctc.py with different classes. Unit tests modified accordingly.
New baseline is the 0.8DO, 6x1024, specaugment. 
Added Lingvo's conformer model (TF2.3) but TPUs seem to have issues with TF2.3.

### CODE REVIEW

**Comment by galv** [comment]
> Added Lingvo's conformer model (TF2.3) but TPUs seem to have issues with TF2.3.

Some thoughts:
- We will need to update to the latest version of lingvo to use TF2.3.
- You may have specified --tf-version 2.2 in the "ctpu up" command on accident.

File: example_ctc_tpu.sh [hunk 1]
```diff
@@ -1,13 +0,0 @@
-export CUDA_VISIBLE_DEVICES=""
```
Comment (galv) [inline example_ctc_tpu.sh:1]: This has a conflict with master. I suggest you just retain the file on master for now rather than deleting it.

File: lingvo/tasks/asr/wpm_16k_librispeech_ascii.vocab [hunk 2]
```diff
@@ -0,0 +1,2555 @@
+<unk>
```
Comment (galv) [inline lingvo/tasks/asr/wpm_16k_librispeech_ascii.vocab]: What was wrong with the original file https://github.com/galv/lingvo-copy/blob/master/lingvo/tasks/asr/wpm_16k_librispeech.vocab? Was the encoding problematic?

File: lingvo/tasks/asr/params/librispeech_ctc.py [hunk 3]
```diff
@@ -104,40 +106,261 @@ def Task(self):
...
     return p
 
   def ProgramSchedule(self):
     return program.SimpleProgramScheduleForTask(
         train_dataset_name='Train',
-        train_steps_per_loop=50,
-        eval_dataset_names=['Dev'],
-        eval_steps_per_loop=5,
+        train_steps_per_loop=500,
+        eval_dataset_names=['Dev', 'Train'],
```
Comment (galv) [inline lingvo/tasks/asr/params/librispeech_ctc.py:137]: Note that I don't expect evaluating on the Train dataset will be very meaningful. We already get log likelihood (loss) from the Train task on the Train dataset.

File: lingvo/tasks/asr/params/librispeech_ctc.py [hunk 4]
```diff
@@ -52,11 +54,11 @@ def _CommonInputParams(self, is_eval):
       p.source_max_length = 1710
       p.bucket_upper_bound = [639, 1062, 1275, 1377, 1449, 1506, 1563, 1710]
 
-    # p.bucket_batch_limit = [96, 48, 48, 48, 48, 48, 48, 48]
-    p.bucket_batch_limit = [2 * x for x in [48, 48, 48, 48, 48, 48, 48, 48]]
-
-    # Assumes ascii_tokenizer.cc. Gross!
-    p.tokenizer.vocab_size = 76
+    # AG TODO: For TPU
+    p.bucket_batch_limit = [48] * 8
```
Comment (galv) [inline lingvo/tasks/asr/params/librispeech_ctc.py:53]: Did you get out-of-memory errors when you used batch size of 96?

File: lingvo/tasks/asr/params/librispeech.py [hunk 5]
```diff
@@ -38,7 +38,9 @@ def _CommonInputParams(self, is_eval):
     # Generated using scripts in lingvo/tasks/asr/tools.
     p.file_datasource = datasource.PrefixedDataSource.Params()
     p.file_datasource.file_type = 'tfrecord'
+    # AG TODO: chang this local path to gs path for orig data (this is for testing only)
```
Comment (galv) [inline lingvo/tasks/asr/params/librispeech.py]: Please do that for the code you check in.

File: lingvo/tasks/asr/params/librispeech.py [hunk 6]
```diff
@@ -53,9 +55,11 @@ def _CommonInputParams(self, is_eval):
       p.bucket_upper_bound = [639, 1062, 1275, 1377, 1449, 1506, 1563, 3600]
     else:
       p.source_max_length = 3000
+      # handle varying input sizes
+      # clips with <=639 (?) time units (stesps) fall into bucket 1 and so on
```
Comment (galv) [inline lingvo/tasks/asr/params/librispeech.py]: That is correct.

Please remove these note-to-self comments before merging the PR, by the way.

File: lingvo/tasks/asr/params/librispeech.py [hunk 7]
```diff
@@ -53,9 +55,11 @@ def _CommonInputParams(self, is_eval):
...
       p.bucket_upper_bound = [639, 1062, 1275, 1377, 1449, 1506, 1563, 1710]
 
-    p.bucket_batch_limit = [96, 48, 48, 48, 48, 48, 48, 48]
+    p.bucket_batch_limit = [x//8 for x in [96, 48, 48, 48, 48, 48, 48, 48]]
```
Comment (galv) [inline lingvo/tasks/asr/params/librispeech.py]: What is this change for? Are you still using lingvo/tasks/asr/params/librispeech.py or have you moved onto librispeech_ctc.py?

File: lingvo/tasks/asr/encoder.py [hunk 8]
```diff
@@ -64,9 +64,11 @@ def Params(cls):
     p.Define('input_shape', [None, None, None, None],
```
Comment (galv) [inline lingvo/tasks/asr/encoder.py]: Are you still using your changes to encoder.py or are all of your current models using encoder_v2.py?

File: lingvo/core/ops/assert_kernels.cc [hunk 9]
```diff
@@ -103,10 +103,10 @@ class AssertSameDim0Op : public OpKernel {
 REGISTER_KERNEL_BUILDER(Name("AssertSameDim0").Device(DEVICE_CPU),
                         AssertSameDim0Op);
 
-#if GOOGLE_CUDA
+// #if GOOGLE_CUDA
```
Comment (galv) [inline lingvo/core/ops/assert_kernels.cc]: If it is not necessary, can we revert this change? I dislike modifying core parts of the library, in case we might mess something up later.

File: lingvo/tasks/asr/ctc_model.py [hunk 10]
```diff
@@ -34,212 +38,234 @@ class CTCModel(base_model.BaseTask):
...
+  def _CalculateErrorRates(self, dec_outs_dict, input_batch):
+    sparse_gt_ids = tf.cast(py_utils.SequenceToSparseTensor(
+      input_batch.tgt.labels, input_batch.tgt.paddings), tf.int64)
+
+    gt_seq_lens = py_utils.LengthsFromBitMask(input_batch.tgt.paddings, 1)
+    gt_transcripts = self.input_generator.IdsToStrings(
+      input_batch.tgt.labels, gt_seq_lens)
+
+    # char error rate
+    # AG TODO: This is counting num tokens, not num chars.
```
Comment (galv) [inline lingvo/tasks/asr/ctc_model.py]: ah, yes, I suppose we do want token error rate when we move onto word piece models. I wouldn't worry too much about this right now TBH. Sometimes ASR papers site TER (token error rate) but Word error rate is really the gold standard because it is reflective of what matters for production ASR systems.

File: lingvo/tasks/asr/ctc_model.py [hunk 11]
```diff
@@ -34,212 +38,234 @@ class CTCModel(base_model.BaseTask):
...
+
+    # Some predictions have start and stop tokens predicted, we dont want to include
+    # those in WER calculation
+    hyp_str = tf.strings.regex_replace(hyp_str, '(<unk>)+', '')
+    hyp_str = tf.strings.regex_replace(hyp_str, '(<s>)+', '')
+    hyp_str = tf.strings.regex_replace(hyp_str, '(</s>)+', '')
+    return py_utils.NestedMap(sparse_ids=decoded, transcripts=hyp_str)
+
+  def _CalculateErrorRates(self, dec_outs_dict, input_batch):
+    sparse_gt_ids = tf.cast(py_utils.SequenceToSparseTensor(
```
Comment (galv) [inline lingvo/tasks/asr/ctc_model.py]: Dead code?

File: lingvo/tasks/asr/ctc_model.py [hunk 12]
```diff
@@ -34,212 +38,234 @@ class CTCModel(base_model.BaseTask):
...
-          'src_frames': input_batch_src.src_inputs,
-          'encoder_frames': encoder_outputs.encoded
-      }
+    encoded, padding = self.project_to_vocab_size(encoder_out.encoded, encoder_out.padding)
+    outputs = py_utils.NestedMap(encoded=encoded, padding=padding)
+    return outputs
 
-      return fetches, feeds
+  def _DecodeCTC(self, output_batch):
+    # swap row 0 and row 73 because decoder assumes blank is at 0,
```
Comment (galv) [inline lingvo/tasks/asr/ctc_model.py]: BTW, my new tokenizer always assumes blank is at row 0 because the WFST decoder depends upon that being the case. It's rather silly to have a magic constant like this, but it seems the machine learning world has settled upon 0 being the index for blank.

**Review by galv — COMMENTED** [review]
Definitely a lot of things here. I don't understand all of it, but note that I don't necessarily *need* to, so let's not let that hold us back indefinitely.

I do have a request (which I should do myself as well). Can you install pylint and use this pylintrc file https://github.com/tensorflow/lingvo/issues/228#issuecomment-645034928 to check for errors like unused variables and dead code? I noticed one instance of this with my eyes. Ask me for help if you're not sure how to use it.

File: lingvo/core/py_utils.py [hunk 13]
```diff
@@ -5426,3 +5426,10 @@ def ShardedFilePatternToGlob(file_pattern):
   if shards == '*':
     return f'{path}-?????-of-*'
   return f'{path}-?????-of-{int(shards):05}'
+
+def SequenceToSparseTensor(dense_tensor, padding):
```
Comment (galv) [inline lingvo/core/py_utils.py]: You can probably revert this if yoru one usage of it is really for an unused variable

File: example_ctc_tpu.sh (hunk 1 above)
Comment (anjaligopi) [reply example_ctc_tpu.sh:1]: Yeah right, I intend to clean this one before merging. I realized about the conflict after deleting it.

File: lingvo/tasks/asr/params/librispeech_ctc.py (hunk 4 above)
Comment (anjaligopi) [reply lingvo/tasks/asr/params/librispeech_ctc.py:53]: Yeah, that's right. I can't remember exactly when I started seeing those OOM issues but I did see it one of the hyperparam tuning stages.

File: lingvo/tasks/asr/encoder.py (hunk 8 above)
Comment (anjaligopi) [reply lingvo/tasks/asr/encoder.py]: I have a few classes with encoder (including the 17.9 WER baseline) and most of the newer ones are with encoder v2. Should eventually move everything to v2.

File: lingvo/core/ops/assert_kernels.cc (hunk 9 above)
Comment (anjaligopi) [reply lingvo/core/ops/assert_kernels.cc]: agreed

File: lingvo/tasks/asr/wpm_16k_librispeech_ascii.vocab (hunk 2 above)
Comment (anjaligopi) [reply lingvo/tasks/asr/wpm_16k_librispeech_ascii.vocab]: It had 16k tokens - didnt even fit in memory. I wanted to start with a smaller subset.

File: lingvo/tasks/asr/params/librispeech_ctc.py (hunk 3 above)
Comment (anjaligopi) [reply lingvo/tasks/asr/params/librispeech_ctc.py:137]: Agreed, just wanted to see what the loss is like. It's interesting to see the loss reported by evaler on the train set to be lower than the train loss. Triaining has DO, specaugment but eval doesnot.

File: lingvo/tasks/asr/params/librispeech.py (hunk 5 above)
Comment (anjaligopi) [reply lingvo/tasks/asr/params/librispeech.py]: It's probably a stale comment that I need to get rid of. It's pointing to the GS path.

File: lingvo/tasks/asr/ctc_model.py (hunk 11 above)
Comment (anjaligopi) [reply lingvo/tasks/asr/ctc_model.py]: It's used in line 111. Why do you say it's dead?

File: lingvo/tasks/asr/ctc_model.py (hunk 12 above)
Comment (anjaligopi) [reply lingvo/tasks/asr/ctc_model.py]: Stale comment. No longer hard code the idx of blank token, we read the index of blank token from params. So, it should work with whatever index your tokenizer has.

File: lingvo/core/py_utils.py (hunk 13 above)
Comment (anjaligopi) [reply lingvo/core/py_utils.py]: Oh you are referring to the sparse_gt_ids variable as the dead code in you r previous comment, yes, will fix this one. Will look into pylint.

File: lingvo/tasks/asr/wpm_16k_librispeech_ascii.vocab (hunk 2 above)
Comment (galv) [reply lingvo/tasks/asr/wpm_16k_librispeech_ascii.vocab]: Oh I see, and this one has only 2555 tokens. Do note that the "16k" in the model is somewhat confusing then. The word piece models are probably created by this software: https://github.com/google/sentencepiece so I would recommend removing this file for now. If we want to retry this, it is probably best to generate a tokenizer with a smaller number of tokens.

This OOM is definitely interesting. The RNN-T models described in this paper https://arxiv.org/abs/1811.06621 use a 16k WPM vocabulary size, with a batch size of 4096/64 = 64 on each TPUv3.

File: lingvo/tasks/asr/ctc_model.py (hunk 11 above)
Comment (galv) [reply lingvo/tasks/asr/ctc_model.py]: I mean that sparse_gt_ids is an unused variable, to clarify.

**Comment by anjaligopi** [comment]
Linted and formatted. I've addressed and resolved most of the comments. Lets get this merged! I can make the further changes in a separate PR.

File: tpu_submit.sh [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+# if tpu- in $1
+if [[ "$1" == *"-tpu"* ]]; then
+    LOGDIR="${GS_BASE}/${FLDRDATE}/${CLS}"
+
+    name=$1
+    ip_addr=$(../ctpu st -name ${name} --details | grep "TPU IP" | grep -oP "10.*")
+
+    if [ -z "$ip_addr" ]; then
+        echo "Couldnt find TPU, creating a new one"
+        ../ctpu up -name ${name} -tpu-only -tpu-size v3-8 -tf-version 2.2
```
Comment (galv) [inline tpu_submit.sh]: You don't need to change this, but just a note, this assumings the path of ctpu.

**Comment by galv** [comment]
Please go ahead and merge it!
